// pdf-lib/cjs/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./colors"), exports);
tslib_1.__exportStar(require("./errors"), exports);
tslib_1.__exportStar(require("./objects"), exports);
tslib_1.__exportStar(require("./operations"), exports);
tslib_1.__exportStar(require("./operators"), exports);
tslib_1.__exportStar(require("./rotations"), exports);
tslib_1.__exportStar(require("./sizes"), exports);
tslib_1.__exportStar(require("./PDFPageOptions"), exports);
tslib_1.__exportStar(require("./StandardFonts"), exports);
var PDFDocument_1 = require("./PDFDocument");
exports.PDFDocument = PDFDocument_1.default;
exports.ParseSpeeds = PDFDocument_1.ParseSpeeds;
var PDFFont_1 = require("./PDFFont");
exports.PDFFont = PDFFont_1.default;
var PDFImage_1 = require("./PDFImage");
exports.PDFImage = PDFImage_1.default;
var PDFPage_1 = require("./PDFPage");
exports.PDFPage = PDFPage_1.default;
var PDFEmbeddedPage_1 = require("./PDFEmbeddedPage");
exports.PDFEmbeddedPage = PDFEmbeddedPage_1.default;
